onnxruntime/core/providers/cuda/reduction/reduction_ops.cc

```diff
@@ -438,7 +439,11 @@ Status ReduceComputeCore(CUDAExecutionProvider& cuda_ep, const Tensor& input, Pr
       } else {
         // Reduce max -- Max/Min will output indices data
         CudnnReduceDescriptor reduce_max_desc;
-        ORT_RETURN_IF_ERROR(reduce_max_desc.Set(CUDNN_REDUCE_TENSOR_MAX, cudnn_type_X, CUDNN_REDUCE_TENSOR_NO_INDICES));
+        cudnnDataType_t cudnn_reduce_max_type = cudnn_type_X;
+        if((std::is_same<T, MLFloat16>::value)) {
+            cudnn_reduce_max_type = CUDNN_DATA_FLOAT;
```

Comment:
This is fixing for ReduceMax? 
I thought we have problem for ReduceLogSumExp? 

Also, I am not sure how well cudnn reduction works for half type, Weixing had some concerned with this before. 
Another approach is to cast is to fp32 first and use cudnn reduction in float.

Reply:
Yea this whole function is written in a strange way. But the flow is: 1. get the index space for reducing(this is for keepdim since whether to keep dim or not will change output shape) 2. reduceMax on the input and copy result. 3. perform log->reduceSum->exp... Step 1 and step 2 need to be using the same type, in this case, step 1 was using a different type as step 2, it was complaining about type mismatch.. But the runtime type of the actual computation in step 3 is all done in float.

Reply:
I see. ReduceMax is used to avoid numerical problem of doing exp direction on x.